### DISCUSSION TITLE
RFC: Angular Split with Snap feature

### DISCUSSION THREAD

**Discussion by Jefiozie** [bertrandg/angular-split#458] [RFCs]
### **Background**

In our recent "modernize Angular split" PR, we encountered significant issues with the existing collapse/expand feature, particularly when collapsing to a non-zero value. This behavior is not intuitive for a split library. Consequently, we propose developing a true snap feature from the ground up, leveraging the enhancements from the "modernize" PR for easier maintenance.

### **Community Input**

We seek feedback from the community on the following:

- How do you currently use the collapse/expand feature?

- Would a snap feature better serve your needs?

- What are your use cases for collapsing to a non-zero size?

- Will this proposal meet your requirements?

### **High-Level Proposal**

Our snap feature aims to:

1. Function independently without requiring state synchronization by the developer.

2. Provide easy state-saving mechanisms when needed.

3. Support both visible and `@if`based snapping.

4. Facilitate the restoration of a snapped area to its pre-snap size (likely requiring state management).

5. Adhere to `maxSize` and `lockSize` constraints.

6. Allow configuration of snap feature enablement.

The proposed snap feature will emulate the snapping behavior of VS Code. When dragging an area with a defined `minSize`, it will snap when the pointer reaches half of the area size. At this point, the gutter remains visible, implying the area has zero size. If the pointer moves past half the area minSize towards expansion, the area snaps back to the minSize and grows from there. Releasing the pointer at the snap position commits the snap, similar to current size changes.

Upon snapping, the area's visibility will automatically be set to false internally. Developers can choose to update their visible state or remove the area using `@if`.

**Notable changes**

The current event `SplitGutterInteractionEvent `will change in the case of drag end to a more specific event (`SplitGutterDragEndEvent`) containing two important properties to describe the snap.

1. `snapAwareSizes `- As the area visibility changed to false, the regular event’s sizes array will include less elements than before. This new property will describe the new sizes array while maintaining the snapped area size. In every drag end event with no snap involved - the sizes array and the `snapAwareSizes `will be the same.

2. `snappedAreaIndex `- This will be the index of the area that snapped. In every drag end event with no snap involved - the `snappedAreaIndex `will be -1.

**For example:**

Considering a percent based split with the following area sizes: `[40, 20, 40]`, the user snaps the first area. At this point the sizes array on the drag end event will be` [60, 40]` as the first area snapped. But, `snapAwareSizes `will be `[40, 60, 40] `to allow the same index access and the old area size if needed and `snappedAreaIndex `will be` 0`.

### **Request for Feedback**

We believe this behavior will provide a good alternative to the old collapse feature and will allow the developer the freedom to choose how to handle the snap that occured.

We value your insights and feedback on this proposal. Please share your experiences and thoughts to help us refine and implement a feature that best meets the community's needs.

**Comment by ranitg** [comment]
What should I use instead of collapse/expand? I used them to have something like the vscode splitter. I can resize the sections and also click on the headers and then it would collapse/expand. It doesn't collapse to zero size but to the size of the sections header, and then expands back to the its previous size. Both can be collapsed at the same time. Also is ti possible to have only two sections?

**Reply by grafanauser** [reply]
same here. We use it to collapse and expand sidebars via a button click. 
- The collapsed sidebars are small and not resizable. 
- The expanded sizebars are bigger and resizable 
- an expanded sidebar shows different content than the sidebar in collapsed state. 
- Grabbing the sidebar instead with the mouse to collapse and expand it is not an option. This kind of interaction is far too hidden for a user to find out about in this case. Collapse()/expand(40) must be called with a button click event instead. Yes the expanded version is resizeable, but the user does not rely on knowing that this is possible. So being able to resize the expanded sidebar is for more advanced users but it is still required.

**Reply by Harpush** [reply]
@grafanauser  Except the different size of the gutter this can be achieved with:
1. `lockSize` allows to lock the area and disable resizing
2. `size` binding allows changing the size
3. `*asSplitGutter` allows customizing the gutter appearance 

For your case the collapse and expand functionality can be achieved with the current API. The only "missing" part is making the gutter smaller. You can however alter the styles to indicate the gutter is currently disabled.

**Reply by grafanauser** [reply]
@Harpush 
`size` is a component @input. How can i read the current size when a user dragged the gutter for resizing? When I collapse the sidebar, I need to store the old expanded size to be able to restore the size again when the sidebar is expanded again. That worked out of the box in the old angular-split version. The new version probably needs an `@Output sizeChange`?

**Reply by Harpush** [reply]
First of all in the old version it worked as long as you didn't need to persist the state if I remember correctly.
You have the dragEnd to get the size.

**Reply by grafanauser** [reply]
Thanks, that worked. For people trying to achieve the same:
- you need to bind to as-split-area `size` and `lockSize`. Set those in the click handler of the expand/collapse button. store the expanded/collapsed sidebar state in the component too (e.g. `isCollapsed`).
- you need to bind a variable to as-split `gutterSize` if you want the collapsed sidebar to be a thin line while the expanded sidebar is thicker for mouse grap resizing.
- to restore the size of a previously by the user resized expanded sidebar, you need to bind to as-split `(dragEnd)="onDragEnd($event)`. In `onDragEnd` store the `SplitAreaSize` from `e.sizes` that you need when the sidebar is in the expanded state. `onDragEnd(e: SplitGutterInteractionEvent) { [this.oldExpandedSize] = e.sizes; }`
- for the collapsed state you need to add `@if(isCollapsed) { <div *asSplitGutter class="custom-split-gutter"></div> }` with `.custom-split-gutter {  width: 100%; height: 100%; cursor: default; }`. This will remove the drag cursor and hide the dotted drag-icon on the gutter itself.

**Reply by grafanauser** [reply]
the warning spam is annoying:
> split-area.component.ts:136 as-split: lockSize overwrites maxSize/minSize

**Reply by Harpush** [reply]
1. Only happens in dev mode.
2. Why would you put both min/max and lock? Generally it should be either of them

**Reply by grafanauser** [reply]
- lock is for collapsed sidebar. This is set/reset programatically via a binding.
- min/max is required to limit resizeability of expanded sidebar. This is simply set on the <as-split-area minSize=..> in the html template.

example picture with sidebar without content, just expand/collapse button and split line visible
![image](https://github.com/user-attachments/assets/f946edef-6a6f-4025-b21f-41a995e405cc)

**Reply by Harpush** [reply]
Well min/max can be set to `*` or `undefined` when `lockSize` is true to avoid the warning.
Anyway if you think the warning shouldn't be there you are free to make an issue on that with a real case scenario why it makes sense.

**Comment by er1z** [comment]
Is this topic still alive?

**Reply by Harpush** [reply]
There wasn't much interest in this feature... But if there will be it is still under consideration.

**Reply by er1z** [reply]
It would be nice to have some sort of snapping. Besides, is there any way to have work around? I'm pretty new to this library and I'm doing something like partition alignement but want to have round values in result.

**Reply by Harpush** [reply]
Not sure what partition alignment means. But snapping like vscode is currently not supported and without an obvious workaround

**Reply by er1z** [reply]
Ahh, so I guess we understand `snapping` differently.

To snap I meant something like this:
<img width="1000" height="750" alt="image" src="https://github.com/user-attachments/assets/b8e000ac-5163-41a0-96bd-da2248a9c8eb" />

E.g. to allow resizing only by multiplies of X, let's say by 100, not by one.

**Reply by Harpush** [reply]
You might need some math here but I think `gutterStep` can answer your needs

**Reply by er1z** [reply]
Perfect! I wish it also supported percentages but I'm still closer. Thank you for an advice and sorry for messing up the topic.

**Reply by Harpush** [reply]
You can open an issue concerning percent support. One can say that supporting percent in gutter step is similar to percent in min/max... Though I haven't thought about enough to be certain
